internal/server/filesystem/dump_adapters.go

```diff
@@ -124,3 +125,69 @@ func PrepareDumpDir(next types.Dump) types.Dump {
 		return exited, nil
 	}
 }
+
+// This adapter ensures the specified dump dir exists and is writable.
+// Creates a unique directory within this directory for the dump.
+// Updates the CRIU server to use this newly created directory.
+// Compresses the dump directory post-dump, based on a compression format:
+//   - "none" does not compress the dump directory
+//   - "tar" creates a tarball of the dump directory
+//   - "gzip" creates a gzipped tarball of the dump directory
+//   - "lz4" creates an lz4-compressed tarball of the dump directory
+func PrepareDumpVMDir(next types.DumpVM) types.DumpVM {
+	return func(ctx context.Context, opts types.Opts, resp *daemon.DumpVMResp, req *daemon.DumpVMReq) (exited chan int, err error) {
+		dir := req.GetDir()
+		compression := config.Global.Checkpoint.Compression
+
+		// Check if the provided dir exists
+		if _, err := os.Stat(dir); os.IsNotExist(err) {
+			return nil, status.Errorf(codes.InvalidArgument, "dump dir does not exist: %s", dir)
+		}
+
+		f, err := os.Open(dir)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to open dump dir: %v", err)
+		}
+		defer f.Close()
+
+		// Setup dump fs that can be used by future adapters to directly read write/extra files
+		// to the dump directory
+		opts.DumpFs = afero.NewBasePathFs(afero.NewOsFs(), dir)
+
+		if req.Type == "cloud-hypervisor" {
```

Comment:
This logic can thus end up in the CLH plugin as a new adapter that simply modifies the `req.Dir`.